src/accountsdb/buffer_pool.zig

```
@@ -0,0 +1,1074 @@
+        const rc = try allocator.alignedAlloc(
+            sig.sync.ReferenceCounter,
+            std.mem.page_size,
+            num_frames,
+        );
+        errdefer allocator.free(rc);
+        @memset(rc, .{ .state = .{ .raw = 0 } });
+
+        const key = try allocator.alignedAlloc(FileIdFileOffset, std.mem.page_size, num_frames);
+        errdefer allocator.free(key);
+        @memset(key, .{ .file_id = FileId.fromInt(0), .file_offset = 0 });
+
+        const freq = try allocator.alignedAlloc(u2, std.mem.page_size, num_frames);
+        errdefer allocator.free(freq);
+        @memset(freq, 0);
+
+        const in_queue = try allocator.alignedAlloc(InQueue, std.mem.page_size, num_frames);
+        errdefer allocator.free(in_queue);
+        @memset(in_queue, .none);
+
+        const size = try allocator.alignedAlloc(FrameOffset, std.mem.page_size, num_frames);
+        errdefer allocator.free(size);
+        @memset(size, 0);
+
+        return .{
+            .rc = rc,
+            .key = key,
+            .freq = freq,
+            .in_queue = in_queue,
+            .size = size,
+        };
+    }
+
+    fn deinit(self: *FramesMetadata, allocator: std.mem.Allocator) void {
+        // NOTE: this check itself is racy, but should never happen
+        for (self.rc) |*rc| {
+            if (rc.isAlive()) {
+                @panic("BufferPool deinitialised with alive handles");
+            }
+        }
+        allocator.free(self.rc);
+        allocator.free(self.key);
+        allocator.free(self.freq);
+        allocator.free(self.in_queue);
+        allocator.free(self.size);
+        self.* = undefined;
+    }
+
+    // to be called on the eviction of a frame
+    // should never be called on a frame with rc>0
+    // TODO: this should *all* be atomic (!)
+    fn resetFrame(self: FramesMetadata, index: FrameIndex) void {
+        if (self.rc[index].isAlive()) unreachable;
```

should just throw an error -- same comment as other reachable unreachable

---

downgraded to an error. https://github.com/Syndica/sig/pull/368/commits/1c8f4bdcfe4295dac4f14acabd0ae3af413c169e#diff-9334997573b7124ee92d8ed1efaa5a70b34f8531988349743900d1cf39014569L494-R522

however this error should never be reachable by using the public api incorrectly